src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs

```diff
@@ -58,33 +69,77 @@ public static unsafe int HashData(string hashAlgorithmId, ReadOnlySpan<byte> sou
+            {
+                hashSize = default;
+
+                Interop.BCrypt.BCryptAlgPseudoHandle algHandle;
+                int digestSizeInBytes;
+
+                if (hashAlgorithmId == HashAlgorithmNames.MD5)
                 {
-                    try
-                    {
-                        fixed (byte* pSource = source)
-                        fixed (byte* pDestination = destination)
-                        {
-                            ntStatus = Interop.BCrypt.BCryptHash(cachedAlgorithmHandle, null, 0, pSource, source.Length, pDestination, hashSize);
+                    algHandle = Interop.BCrypt.BCryptAlgPseudoHandle.BCRYPT_MD5_ALG_HANDLE;
+                    digestSizeInBytes = 128 / 8;
+                }
+                else if (hashAlgorithmId == HashAlgorithmNames.SHA1)
+                {
+                    algHandle = Interop.BCrypt.BCryptAlgPseudoHandle.BCRYPT_SHA1_ALG_HANDLE;
+                    digestSizeInBytes = 160 / 8;
+                }
+                else if (hashAlgorithmId == HashAlgorithmNames.SHA256)
+                {
+                    algHandle = Interop.BCrypt.BCryptAlgPseudoHandle.BCRYPT_SHA256_ALG_HANDLE;
+                    digestSizeInBytes = 256 / 8;
+                }
+                else if (hashAlgorithmId == HashAlgorithmNames.SHA384)
+                {
+                    algHandle = Interop.BCrypt.BCryptAlgPseudoHandle.BCRYPT_SHA384_ALG_HANDLE;
+                    digestSizeInBytes = 384 / 8;
+                }
+                else if (hashAlgorithmId == HashAlgorithmNames.SHA512)
+                {
+                    algHandle = Interop.BCrypt.BCryptAlgPseudoHandle.BCRYPT_SHA512_ALG_HANDLE;
+                    digestSizeInBytes = 512 / 8;
+                }
+                else
+                {
+                    Debug.Fail("Unknown hash algorithm.");
+                    return false;
+                }
 
-                            if (ntStatus != NTSTATUS.STATUS_SUCCESS)
-                            {
-                                throw Interop.BCrypt.CreateCryptographicException(ntStatus);
-                            }
-                        }
+                if (destination.Length < digestSizeInBytes)
+                {
+                    Debug.Fail("Caller should have checked length.");
+                    return false;
+                }
 
-                        return hashSize;
+                NTSTATUS ntStatus;
+                fixed (byte* pSrc = &MemoryMarshal.GetReference(source))
+                fixed (byte* pDest = &MemoryMarshal.GetReference(destination))
+                {
+                    try
+                    {
+                        ntStatus = Interop.BCrypt.BCryptHash((uint)algHandle, null /* pbSecret */, 0 /* cbSecret */, pSrc, source.Length, pDest, digestSizeInBytes);
```

Comment:
The comments could be named for parameters, too, I think:

```suggestion
                        ntStatus = Interop.BCrypt.BCryptHash((uint)algHandle, pbSecret: null, cbSecret: 0, pSrc, source.Length, pDest, digestSizeInBytes);
```